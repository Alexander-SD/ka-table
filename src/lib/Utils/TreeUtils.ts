
import { getValueByField } from './DataUtils';

export const treeGroupMark = {};
export const treeDataMark = {};


export const getExpandedParents = (treeData: any[], rowKeyField: any): any[][] =>
  treeData
    .filter((item) => item.treeGroupMark === treeGroupMark)
    .map((item) => getValueByField(item.rowData, rowKeyField));

const getItemStructure = (
  item: any,
  dataHash: any,
  rowKeyField: any,
  deep: number = 0): any[] => {
    const children = dataHash[getValueByField(item, rowKeyField)];
    if (!children){
      return [{ treeDataMark, rowData: item, deep: deep + 1 }];
    }
    const result = [{ treeGroupMark, rowData: item, deep }];
    children.forEach((c: any) => {
      const childrenData = getItemStructure(c, dataHash, rowKeyField, deep + 1);
      result.push(...childrenData);
    });
    return result;
};

export const getTreeData = (
  data: any[],
  rowKeyField: any,
  parentRowKeyField: any,
  parentsExpanded?: any[][]): any[] => {
    const dataHash: any = {};
    const rootElements: any[] = [];
    data.forEach(d => {
      const parentRowKeyValue = getValueByField(d, parentRowKeyField) ?? undefined;
      if (!parentRowKeyValue){
        rootElements.push(d);
        return;
      }
      if (!dataHash[parentRowKeyValue]){
        dataHash[parentRowKeyValue] = [];
      }
      if (!parentsExpanded || parentsExpanded.includes(parentRowKeyValue)) {
        dataHash[parentRowKeyValue].push(d);
      }
    });
    const newData: any[] = [];
    rootElements.forEach(d => {
      newData.push(...getItemStructure(d, dataHash, rowKeyField));
    });
    return newData;
};

