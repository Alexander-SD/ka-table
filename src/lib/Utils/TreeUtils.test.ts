import { getExpandedParents, getTreeData, treeDataMark, treeGroupMark } from './TreeUtils';

describe('TreeUtils', () => {
  describe('getTreeData', () => {
    it('default', () => {
      const data = [
        { parentId: null, id: 1, name: 'level 1 - 1', orderCount: 5 },
        { parentId: 1, id: 2, name: '1 - Data 1', orderCount: 2 },
        { parentId: 1, id: 3, name: '1 - Data 2', orderCount: 3 },
        { parentId: null, id: 4, name: 'Level 1 - 2', orderCount: 7 },
        { parentId: 4, id: 5, name: '2 - Data 1', orderCount: 2 },
        { parentId: 4, id: 6, name: '2 - Data 2', orderCount: 5 },
      ];
      const treeData = getTreeData(data, 'id', 'parentId', undefined);
      expect(treeData).toMatchSnapshot();
    });
    it('2 levels', () => {
      const data = [
        { parentId: null, id: 1, name: 'level 1 - 1', orderCount: 5 },
        { parentId: 1, id: 7, name: 'level 2 - 1', orderCount: 5 },
        { parentId: 7, id: 2, name: '1 - 1 - Data 1', orderCount: 2 },
        { parentId: 7, id: 10, name: '1 - 1 - Data 2', orderCount: 2 },
        { parentId: 8, id: 3, name: '1 - 2 - Data 1', orderCount: 3 },
        { parentId: 1, id: 8, name: '1 - level 2 - 2', orderCount: 5 },
        { parentId: null, id: 4, name: 'Level 1 - 2', orderCount: 7 },
        { parentId: 4, id: 5, name: '2 - Data 1', orderCount: 2 },
        { parentId: 4, id: 9, name: '2 - level 2 - 2', orderCount: 5 },
        { parentId: 9, id: 6, name: '2 - 2 - Data 1', orderCount: 5 },
      ];
      const treeData = getTreeData(data, 'id', 'parentId', undefined);
      expect(treeData).toMatchSnapshot();
    });

    it('3 levels', () => {
      const data = [
        { parentId: null, id: 1, name: 'level 1 - 1', orderCount: 5 },
        { parentId: 1, id: 7, name: 'level 2 - 1', orderCount: 5 },
        { parentId: 7, id: 2, name: 'level 3 - 1', orderCount: 2 },
        { parentId: 2, id: 10, name: '1 - 1 - 1 - Data 1', orderCount: 2 },
        { parentId: 1, id: 3, name: '1 - Data 1', orderCount: 3 },
      ];
      const treeData = getTreeData(data, 'id', 'parentId', undefined);
      expect(treeData).toMatchSnapshot();
    });
  });

  describe('getExpandedParents', () => {
    it('default', () => {
      const data = [
        { treeGroupMark, rowData: { parentId: null, id: 1, name: 'level 1', orderCount: 5 }},
        { treeGroupMark, rowData: { parentId: 1, id: 2, name: 'level 2', orderCount: 2 }},
        { treeDataMark, rowData: { parentId: 2, id: 3, name: 'Data 1', orderCount: 2 }}
      ];
      const expandedParents = getExpandedParents(data, 'id');
      expect(expandedParents).toEqual([1, 2]);
    });
  });
});
