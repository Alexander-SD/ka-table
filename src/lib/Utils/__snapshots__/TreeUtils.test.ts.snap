// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TreeUtils getTreeData 2 levels 1`] = `
Array [
  Object {
    "deep": 0,
    "rowData": Object {
      "id": 1,
      "name": "level 1 - 1",
      "orderCount": 5,
      "treeGroupId": null,
    },
    "treeGroupMark": Object {},
  },
  Object {
    "deep": 1,
    "rowData": Object {
      "id": 7,
      "name": "level 2 - 1",
      "orderCount": 5,
      "treeGroupId": 1,
    },
    "treeGroupMark": Object {},
  },
  Object {
    "deep": 3,
    "rowData": Object {
      "id": 2,
      "name": "1 - 1 - Data 1",
      "orderCount": 2,
      "treeGroupId": 7,
    },
    "treeDataMark": Object {},
  },
  Object {
    "deep": 3,
    "rowData": Object {
      "id": 10,
      "name": "1 - 1 - Data 2",
      "orderCount": 2,
      "treeGroupId": 7,
    },
    "treeDataMark": Object {},
  },
  Object {
    "deep": 1,
    "rowData": Object {
      "id": 8,
      "name": "1 - level 2 - 2",
      "orderCount": 5,
      "treeGroupId": 1,
    },
    "treeGroupMark": Object {},
  },
  Object {
    "deep": 3,
    "rowData": Object {
      "id": 3,
      "name": "1 - 2 - Data 1",
      "orderCount": 3,
      "treeGroupId": 8,
    },
    "treeDataMark": Object {},
  },
  Object {
    "deep": 0,
    "rowData": Object {
      "id": 4,
      "name": "Level 1 - 2",
      "orderCount": 7,
      "treeGroupId": null,
    },
    "treeGroupMark": Object {},
  },
  Object {
    "deep": 2,
    "rowData": Object {
      "id": 5,
      "name": "2 - Data 1",
      "orderCount": 2,
      "treeGroupId": 4,
    },
    "treeDataMark": Object {},
  },
  Object {
    "deep": 1,
    "rowData": Object {
      "id": 9,
      "name": "2 - level 2 - 2",
      "orderCount": 5,
      "treeGroupId": 4,
    },
    "treeGroupMark": Object {},
  },
  Object {
    "deep": 3,
    "rowData": Object {
      "id": 6,
      "name": "2 - 2 - Data 1",
      "orderCount": 5,
      "treeGroupId": 9,
    },
    "treeDataMark": Object {},
  },
]
`;

exports[`TreeUtils getTreeData 3 levels 1`] = `
Array [
  Object {
    "deep": 0,
    "rowData": Object {
      "id": 1,
      "name": "level 1 - 1",
      "orderCount": 5,
      "treeGroupId": null,
    },
    "treeGroupMark": Object {},
  },
  Object {
    "deep": 1,
    "rowData": Object {
      "id": 7,
      "name": "level 2 - 1",
      "orderCount": 5,
      "treeGroupId": 1,
    },
    "treeGroupMark": Object {},
  },
  Object {
    "deep": 2,
    "rowData": Object {
      "id": 2,
      "name": "level 3 - 1",
      "orderCount": 2,
      "treeGroupId": 7,
    },
    "treeGroupMark": Object {},
  },
  Object {
    "deep": 4,
    "rowData": Object {
      "id": 10,
      "name": "1 - 1 - 1 - Data 1",
      "orderCount": 2,
      "treeGroupId": 2,
    },
    "treeDataMark": Object {},
  },
  Object {
    "deep": 2,
    "rowData": Object {
      "id": 3,
      "name": "1 - Data 1",
      "orderCount": 3,
      "treeGroupId": 1,
    },
    "treeDataMark": Object {},
  },
]
`;

exports[`TreeUtils getTreeData default 1`] = `
Array [
  Object {
    "deep": 0,
    "rowData": Object {
      "id": 1,
      "name": "level 1 - 1",
      "orderCount": 5,
      "treeGroupId": null,
    },
    "treeGroupMark": Object {},
  },
  Object {
    "deep": 2,
    "rowData": Object {
      "id": 2,
      "name": "1 - Data 1",
      "orderCount": 2,
      "treeGroupId": 1,
    },
    "treeDataMark": Object {},
  },
  Object {
    "deep": 2,
    "rowData": Object {
      "id": 3,
      "name": "1 - Data 2",
      "orderCount": 3,
      "treeGroupId": 1,
    },
    "treeDataMark": Object {},
  },
  Object {
    "deep": 0,
    "rowData": Object {
      "id": 4,
      "name": "Level 1 - 2",
      "orderCount": 7,
      "treeGroupId": null,
    },
    "treeGroupMark": Object {},
  },
  Object {
    "deep": 2,
    "rowData": Object {
      "id": 5,
      "name": "2 - Data 1",
      "orderCount": 2,
      "treeGroupId": 4,
    },
    "treeDataMark": Object {},
  },
  Object {
    "deep": 2,
    "rowData": Object {
      "id": 6,
      "name": "2 - Data 2",
      "orderCount": 5,
      "treeGroupId": 4,
    },
    "treeDataMark": Object {},
  },
]
`;
